File: VB/Controllers/HomeController.vb
Imports System.IO
Imports System.Web
Imports System.Web.Mvc
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraPrintingLinks
Imports DevExpress.Web.Mvc
Imports DevExpress.Web.ASPxGridView
Imports System.Data
Imports DevExpress.XtraReports.UI
Imports System.Collections.Generic
Imports DevExpress.Data.Filtering
Imports VB.WebReportRuntime


Namespace CS.Controllers
    Public Class HomeController
        Inherits Controller
        Public Function Index() As ActionResult
            Return View()
        End Function

        Public Function DataBindingToLargeDatabasePartial() As ActionResult
            Return PartialView("GridViewPartial")
        End Function

        Public Function ExportTo(filterString As String) As ActionResult
            Dim filterCriteria As CriteriaOperator = CriteriaOperator.Parse(filterString)
            Dim products As IList(Of Product) = LargeDatabaseDataProvider.GetProductsByFilter(filterCriteria)

            Dim report As New XtraReport1()
            report.DataSource = products
            ExportReport(report, "largeData", True, "Xlsx")
            Return PartialView("GridViewPartial")
        End Function

        Public Sub ExportReport(report As XtraReport, fileName As String, saveAsFile As Boolean, fileFormat As String)
            Using stream As New MemoryStream()
                report.ExportToXlsx(stream)

                Dim disposition As String = If(saveAsFile, "attachment", "inline")

                Response.Clear()
                Response.Buffer = False
                Response.AppendHeader("Content-Type", String.Format("application/{0}", fileFormat))
                Response.AppendHeader("Content-Transfer-Encoding", "binary")
                Response.AppendHeader("Content-Disposition", String.Format("{0}; filename={1}.{2}", disposition, HttpUtility.UrlEncode(fileName).Replace("+", "%20"), fileFormat))
                Response.BinaryWrite(stream.ToArray())
                Response.[End]()
            End Using
        End Sub

    End Class

End Namespace

File: CS/Controllers/HomeController.cs
using System;
using System.IO;
using System.Web;
using System.Web.Mvc;
using DevExpress.XtraPrinting;
using DevExpress.XtraPrintingLinks;
using DevExpress.Web.Mvc;
using DevExpress.Web.ASPxGridView;
using System.Data;
using WebReportRuntime;
using DevExpress.XtraReports.UI;
using System.Collections.Generic;
using CS.Models;
using DevExpress.Data.Filtering;


namespace CS.Controllers {
    public class HomeController : Controller {
        public ActionResult Index() {
            return View();
        }

        public ActionResult DataBindingToLargeDatabasePartial() {
            return PartialView("GridViewPartial");
        }

        public ActionResult ExportTo(string filterString) {
            CriteriaOperator filterCriteria = CriteriaOperator.Parse(filterString);
            IList<Product> products = LargeDatabaseDataProvider.GetProductsByFilter(filterCriteria);

            XtraReport1 report = new XtraReport1();
            report.DataSource = products;
            ExportReport(report, "largeData", true, "Xlsx");
            return PartialView("GridViewPartial");
        }

        public void ExportReport(XtraReport report, string fileName, bool saveAsFile, string fileFormat) {
            using (MemoryStream stream = new MemoryStream()) {
                report.ExportToXlsx(stream);

                string disposition = saveAsFile ? "attachment" : "inline";

                Response.Clear();
                Response.Buffer = false;
                Response.AppendHeader("Content-Type", string.Format("application/{0}", fileFormat));
                Response.AppendHeader("Content-Transfer-Encoding", "binary");
                Response.AppendHeader("Content-Disposition", string.Format("{0}; filename={1}.{2}", disposition, HttpUtility.UrlEncode(fileName).Replace("+", "%20"), fileFormat));
                Response.BinaryWrite(stream.ToArray());
                Response.End();
            }
        }

    }

}
